Add indicator for metadata changes to Save Panel when reviewing modified entities

artemiomorales opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This is a first iteration of adding metadata to the Save Panel when reviewing modified entities.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Begins addressing https://github.com/WordPress/gutenberg/issues/61405
Related to https://github.com/WordPress/gutenberg/pull/61753
We want to indicate to the user when they are modifying the value of custom fields.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Overrides existing logic to determine whether to show the Save Panel, while also showing that metadata has changed.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
<details>
<summary>1. Register post meta by adding this snippet to your theme's functions.php</summary>

```php
add_action( 'init', 'test_block_bindings' );

function test_block_bindings() {
	register_meta(
		'post',
		'text_field',
		array(
			'show_in_rest'      => true,
			'single'            => true,
			'type'              => 'string',
			'default'           => 'default text value',
		)
	);
}
```

</details>


<details>

<summary>2. Add a paragraph block bound to the custom field using the Code Editor</summary>

```
<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"text_field"}}}}} -->
<p>Paragraph content</p>
<!-- /wp:paragraph -->
```

</details>
 
3. Add a normal paragraph block and modify it.
4. Run the following code in your browser console:
```
const meta = wp.data.select('core/editor').getEditedPostAttribute('meta');
meta['text_field'] = 'testing change';
wp.data.dispatch('core/editor').editPost( meta );
```
5. Press the Save button.
6. Verify that the Save Panel appears and shows that metadata has changed.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

<img width="889" alt="333905890-a40e6c2c-7918-42ce-9f6b-69047fe86971" src="https://github.com/WordPress/gutenberg/assets/5360536/fb63ad5f-6412-48fe-88c7-0835b4deb2b9">

artemiomorales committed:
Add rough prototype showing meta in entities panel

artemiomorales commented:
With this PR, so far I've aimed to just explore the relevant areas we'd likely need to explore to get this functionality working. Some caveats with it right now:
- It creates new functions in the stores that are probably unnecessary rather than using existing functionality
- It currently shows all of the metadata rather than just what's been modified
- How it parses that data is very rough

Potentially this can be a starting point though as we look to get this working.

jasmussen commented:
Nice work. Despite following the steps I wasn't able to make the paragraph editable, so I wasn't able to test this myself. That said, from the screenshot, this _seems_ like a very useful disclosure for people who are using this feature, and an important addon to any other affordances we make. Note I did see some errors with the site editor loading, as part of this draft, just in case that's not on my end.

These are all grouped under "Posts", yet all of these are metadata for one post, correct? If yes, then we should group them under under each page item. In the future you could see multiple pages edited in one session, and published together. The following explores collapsibles, icons, separators, I'd love some @WordPress/gutenberg-design input on those pieces.

<img width="1638" alt="Collapsibles" src="https://github.com/WordPress/gutenberg/assets/1204802/302bf4e4-6133-4c60-9d75-3c037402f5af">

[Figma](https://www.figma.com/design/W2SO7yJG09O65q7JjfXZ4m/Block-connections?node-id=602%3A8369&t=hylo65NmHbRBDyye-1), if inspiration strikes.

But the main challenge to solve is to show that all those checkboxes as being _related to the particular post_, I.e. if you edit multiple posts, or more likely, pages in the site editor, and their bound values, you'd see _a list_ of edited pages. So we need to mainly get that separation right.

jameskoster commented:
> If yes, then we should group them under under each page item

That was my initial reaction upon seeing the mockup. IE shouldn't each field appear within the associated post?

jasmussen commented:
Had a good chat with Artemio around what's feasible for next Friday. So I'd like to clarify a bit what I perceive as important here, and that is, ultimately: to make it clear to a user who's editing bound blocks, that they've done so. The established flow we have for surfacing this, is the pre-publish panel. 

Atomically breaking out every single property, adding icons and collapsibles and indentation as I showed above, that can be a future refinement. The important piece here is the clarity of surfacing: _hey, you edited bound blocks, are you sure?_

And that can be as little as this:

<img width="626" alt="Minimal version" src="https://github.com/WordPress/gutenberg/assets/1204802/76fd380c-bd37-4b60-8153-9eca301f2524">

To me that seems a fine V1 to go with.

artemiomorales committed:
Simplify detection of metadata changes

Removed unnecessary code added in previous commit
and instead modified existing functions to add a flag on
existing dirtyEntityRecord structures to indicate when
metadata changes have been made.

artemiomorales committed:
Remove obsolete code

artemiomorales committed:
Add indicator for bindings to save entities panel

artemiomorales commented:
@jasmussen @SantosGuillamot Ok this is what I've been able to get so far:

<img width="889" alt="Screenshot 2024-05-26 at 9 46 00 PM" src="https://github.com/WordPress/gutenberg/assets/5360536/a40e6c2c-7918-42ce-9f6b-69047fe86971">

I haven't been able to get a checkbox working to granularly handle the editing of the block bindings. I'll keep working on it, but I'm not sure if we'll be able to get that before the WP 6.6 beta. 

- Would just having this indication that bindings changes exist, without a checkbox, be sufficient for the time being?
- Also, I'm unsure how to use the existing classes and components to get the margins shown in the design. Perhaps @richtabor has ideas?

Note that last I checked, the [latest branch for editing the meta fields](https://github.com/WordPress/gutenberg/pull/61753) doesn't work independently, so this PR still has commits from the old version of that branch, which means the code isn't ready to be reviewed, but the UX can be looked at.

jasmussen commented:
The added presence of these items in pre-publish, even if you can't uncheck the bindings and omit them separately, is a crucial part of the flow for awareness of bindings. Which is especially important as the block-toolbar/in-canvas UI has been a trickier piece to get right and might need more time in the oven.

And with the deadline looming, we should start merging. So a 👍 👍 on the design aspect, even if there are some spacings, margins, and line heights we need to follow up with.

gziolo commented on packages/core-data/src/selectors.ts:
```diff
@@ -686,6 +718,9 @@ export const __experimentalGetDirtyEntityRecords = createSelector(
 								entityConfig?.getTitle?.( entityRecord ) || '',
 							name,
 							kind,
+							hasMetaChanges: metadataKeys.includes( primaryKey )
```
Looking at the existing logic, it feels like all the changes required in this selector could get moved to this line. I'm thinking about something like the following:

```js
hasMetaChanges: !! getEntityRecordNonTransientEdits(
	state,
	kind,
	name,
	primaryKey
).?meta,
```

All the conditions should be satisfied because these operations happen only for `primaryKeys` computed in the earlier stage.

gziolo replied:
A related question is whether in the long run, the boolean property is going to be sufficient, or it would be preferred to list all individual modified meta changes. I was about to say that `hasMetaChanges` will become part of the public API, but I see that it's prefixed as experimental so it is less important. Anyway, it's good to answer the question if the boolean value can serve its purpose in the far future. Maybe, it's fine to compute all that locally instead by using `getEntityRecordNonTransientEdits` when it's clear that a given entity has some edits applied.

SantosGuillamot replied:
I was gonna share a similar question. I was wondering if we should abstract it to not include only `metaChanges` but track the different changes made to the post and do with that information whatever we consider.

I'm thinking that in the future we will probably have the possibility to edit "Post Data" like title, description, excerpt... And we might want to do something similar.

artemiomorales replied:
> Anyway, it's good to answer the question if the boolean value can serve its purpose in the far future. 

> I was gonna share a similar question. I was wondering if we should abstract it to not include only `metaChanges` but track the different changes made to the post and do with that information whatever we consider.

Yes I think we'd ideally have a granular listing of all the meta fields that have changed, which means a simple boolean would not be enough in the future, though the boolean is a good starting point.

artemiomorales replied:
I followed @gziolo's advice and simplified the logic 👍

SantosGuillamot commented on packages/editor/src/store/selectors.js:
```diff
@@ -125,7 +125,8 @@ export const hasNonPostEntityChanges = createRegistrySelector(
 				entityRecord.name !== type ||
-				entityRecord.key !== id
+				entityRecord.key !== id ||
+				entityRecord.hasMetaChanges
```
I am not sure if the `hasNonPostEntityChanges` is the proper place to do this. Meta changes are actually part of the post entity, so I find it a bit confusing. I don't know where the selector is used but maybe we need a new one or rename it?

artemiomorales replied:
Ok I introduced a new function called `hasMetaChanges` in order to make this check independent.

SantosGuillamot commented on packages/editor/src/components/entities-saved-states/entity-record-item.js:
```diff
@@ -35,15 +36,31 @@ export default function EntityRecordItem( { record, checked, onChange } ) {
+				/>
+			</PanelRow>
+			{ record.hasMetaChanges && (
+				<PanelRow>
+					<Flex className="entities-saved-states__block-bindings">
+						<Icon
+							className="entities-saved-states__connections-icon"
+							icon={ connection }
+							size={ 24 }
+						/>
+						<span className="entities-saved-states__bindings-text">
+							{ __( 'Block Bindings.' ) }
+						</span>
+					</Flex>
+				</PanelRow>
```
I don't know if the message we show should be "Block Bindings". To me, it seems related to just Post Meta, not Block Bindings. If I am not mistaken, any change made to custom fields (with or without block bindings) will trigger this message, right?

gziolo replied:
I had the same observation. I don't think we track, nor should we even try to do so, whether an edit was triggered by the established Block Binding. It's more about bringing more detailed information about what has been edited, similar to when folks edit the site settings: 

<img width="277" alt="Screenshot 2024-05-27 at 10 47 14" src="https://github.com/WordPress/gutenberg/assets/699132/a5bba9e2-7f3c-4006-8129-c5a7f143f3ae">

artemiomorales replied:
So should we rename the message to "Post Meta" and remove the icon? cc @jasmussen

jasmussen replied:
Maybe, we should flex here to meet the beta deadline to be sure. I do like the icon, and I do think it's important to have some indentation, _or_ space below each post+post-meta chunk, so it's visibly clear that "post meta" is not some single unit you can save globally, it's attached and related to each post.

jameskoster replied:
The tricky thing with 'Post meta' is that the scope is wider than 'Bindings'. 

For instance why would a 'Post meta' item be surfaced in the save panel when a block binding is changed, but not when something like the slug, featured image, or discussion settings are changed?

artemiomorales committed:
Modify message to read 'Post Meta'

artemiomorales committed:
Add store function to check if meta has changed

artemiomorales committed:
Remove obsolete check

artemiomorales commented:
I rebased on `trunk` and made that the target branch in order to make the code easier to review, though that means this PR loses the in-canvas editing capabilities that were in https://github.com/WordPress/gutenberg/pull/61753

With that in mind, it's possible to test this using the store in `wp.data` and the browser console, though be aware that the Save button and entities panel don't appear on first try for some reason 🤔

If using the console to debug, the best bet is to modify existing content in a post, then use the following commands in the console, though you may need to do this twice before it works, likely due to some inconsistency in updating via code versus the UI.

```
const meta = wp.data.select('core/editor').getEditedPostAttribute('meta');
meta['text_field'] = 'testing change';
wp.data.dispatch('core/editor').editPost({ meta });
```

artemiomorales committed:
Simplify logic to check if meta has changed

artemiomorales commented:
Ok so the triggering of the panel doesn't work the first time when following the testing instructions, but it does the second time. As far as I can tell, it works with the proposed in-canvas editing of meta fields in https://github.com/WordPress/gutenberg/pull/61753.

My gut says we shouldn't invest time into making it work in debug mode. Thoughts?

Will continue looking at it to better understand the issue; in the meantime, I made this PR ready for review to continue getting feedback.

jasmussen commented:
> My gut says we shouldn't invest time into making it work in debug mode. Thoughts?

Anything that gets this merged before Friday, IMO.

SantosGuillamot commented on packages/editor/src/store/selectors.js:
```diff
@@ -130,6 +130,22 @@ export const hasNonPostEntityChanges = createRegistrySelector(
+ *
+ * @return {boolean} Whether there are edits or not.
+ */
+export const hasMetaChanges = createRegistrySelector( ( select ) => () => {
+	const dirtyEntityRecords =
+		select( coreStore ).__experimentalGetDirtyEntityRecords();
+	return dirtyEntityRecords.some(
+		( entityRecord ) => entityRecord.hasMetaChanges
+	);
+} );
```
As it is still not 100% clear if we will need this specific selector, or we might want to abstract it to not only track meta changes as suggested [here](https://github.com/WordPress/gutenberg/pull/61811#discussion_r1615603578), maybe it is a good idea to make it private for now?

gziolo replied:
I don't have strong opinions about whether there would be a new selector introduced given that there is `hasNonPostEntityChanges`, which looks very similar. It would definitely help to make it private for the start as it needs more testing (unit tests would help here, too). The function will return `true` if any entity has changed. So it would be valuable to at least narrow down the check to `entityRecord.kind === 'postType'`. In effect, the name would be `hasPostMetaChanges`. It's worth also raising the question, whether we want to observe change to all post types or it should be narrowed down to the currently edited post id.

SantosGuillamot replied:
I was the one suggesting creating a new selector or renaming `hasNonPostEntityChanges`: [link](https://github.com/WordPress/gutenberg/pull/61811#discussion_r1615606135). For me, it felt a bit weird because it actually has `postEntityChanges` when the meta fields change. Having said that, I don't have a strong opinion either.

> So it would be valuable to at least narrow down the check to entityRecord.kind === 'postType'.

This makes sense to me.

> whether we want to observe change to all post types or it should be narrowed down to the currently edited post id.

If we do this, we have to ensure it works properly with the query loop, which includes multiple posts that can be edited.

gziolo commented on packages/core-data/src/selectors.ts:
```diff
@@ -686,6 +687,12 @@ export const __experimentalGetDirtyEntityRecords = createSelector(
 								entityConfig?.getTitle?.( entityRecord ) || '',
 							name,
 							kind,
+							hasMetaChanges: !! getEntityRecordNonTransientEdits(
```
One important aspect to keep in mind is that not all entities define the `meta` field. Entities are defined here:

https://github.com/WordPress/gutenberg/blob/ad3094187dec15e6832c134628561a236cf0b19c/packages/core-data/src/entities.js#L23-L232

Post types and taxonomies are loaded dynamically based on another REST API request. What's important is that it isn't universally applicable. We also cover only post meta as the binding source so it might catch additional unplanned cases.

artemiomorales replied:
Thanks for the note! Is there any actionable step on my end to address this feedback? I'm not sure I follow.

If the entities don't have `meta` defined, then `hasMetaChanges` should just evaluate to `false`, which should result in the desired behavior.

gziolo replied:
We can iterate on it later as this selector is experimental. I'm still not convinced it's a good place to introduce the `hasMetaChanges` field, as it isn't generally applicable to all entities. It's also a derived value that gets computed based on all other data returned from this selector.

SantosGuillamot replied:
I was looking at the different possibilities, and I believe we could use something like [getEntityRecordsEdits](https://github.com/WordPress/gutenberg/blob/c72afdabe993565196aee28b317d001a805f1702/packages/core-data/src/selectors.ts#L762-L771) directly in the new `hasPostMetaChanges` selector we are creating. There, we can get a list of the editedEntities and with `getEntityRecordEdits` (or something similar), check if the meta property is part of that.

That would avoid modifying this. I can take a look at that.

artemiomorales committed:
Update tests

artemiomorales committed:
Make hasMetaChanges selector private

SantosGuillamot commented:
> With that in mind, it's possible to test this using the store in wp.data and the browser console, though be aware that the Save button doesn't activate and the entities panel doesn't appear on first try for some reason 🤔

I am not sure if the problem you are facing is related to the pull request or to the code snippet you are using to test it. It seems that if you don't wrap the meta property when you pass it to `editPost`, it works as expected.

```js
const meta = wp.data.select('core/editor').getEditedPostAttribute('meta');
meta['text_field'] = 'testing change';
wp.data.dispatch('core/editor').editPost( meta );
```

Apart from that, I also tested it using `editEntityRecord` directly, as we are doing in [the post meta PR](https://github.com/WordPress/gutenberg/blob/75dc591707ceb0aec666176ad578118214c79d21/packages/editor/src/bindings/post-meta.js#L28-L34), and it seems to be working fine as well.

gziolo commented:
I observed similar changes in the post editor when editing the post meta set in the connected block by following the instructions shared in the PR's description:


https://github.com/WordPress/gutenberg/assets/699132/50e45a82-c7d2-4633-9c71-2dd3ade4e53e

Have you considered integrating the information about the modified post meta in the pre-publish check as another collapsed panel next to suggestions related to tags and categories?

<img width="1289" alt="Screenshot 2024-05-29 at 15 35 25" src="https://github.com/WordPress/gutenberg/assets/699132/ad6c39a2-e222-4540-8feb-fb8d5cba5311">

artemiomorales commented:
> I am not sure if the problem you are facing is related to the pull request or to the code snippet you are using to test it. It seems that if you don't wrap the meta property when you pass it to editPost, it works as expected.

@SantosGuillamot Ok this PR is no longer causing the entities panel to appear for me when testing using the snippet, with or without the curly braces around `meta`. I'm not sure why.

Additionally, I was noticing inconsistencies wherein the Save button would activate after using the snippet, but only in Twenty Twenty Three and not in Twenty Twenty Four or Twenty Twenty Two, though I've just rebased on trunk, which seems to have fixed that issue.

Overall, I'm seeing inconsistencies here on my end and am trying to figure out the cause.

artemiomorales commented:
> Have you considered integrating the information about the modified post meta in the pre-publish check as another collapsed panel next to suggestions related to tags and categories?

@gziolo Thanks for looking at this! Agreed that adding the post meta check in the pre-publish check would be a good move. Will be looking at that as I try to clear up the inconsistencies on my end around the functionality.

SantosGuillamot commented:
> Ok this PR is no longer causing the entities panel to appear for me when testing using the snippet, with or without the curly braces around meta. I'm not sure why.

I'm using `editEntityRecord` to test it, and it seems to work fine 🤷 

```js
wp.data.dispatch('core').editEntityRecord( 'postType', 'page', 410, { meta: { text_field: "New value" } } )
```

What challenges we need to solve to get this pull request merged, then? Some of them that come to my mind after reading the discussion.
* Integrate it with the pre-publish check.
* Check the "Site updated" message.
* Ensure it works well with footnotes (maybe we can skip it manually for now).
* Move any changes to the store to private.
* Add some e2e tests.

Then, we would need to follow-up and discuss the appropriate place for the logic added and how the ideal UX would be.

Anything else @artemiomorales ?

artemiomorales commented:
> I'm using editEntityRecord to test it, and it seems to work fine 🤷

@SantosGuillamot If it works well for you and others, maybe it's a problem with my environments or some other detail I'm missing. In any case, I've moved on and am working on the issues you've described.

Your summary seems right to me:

>- Integrate it with the pre-publish check.
>- Check the "Site updated" message.

I'm currently working on these. 

>- Ensure it works well with footnotes (maybe we can skip it manually for now).

I can look into skipping it manually if that seems like a good solution.

>- Move any changes to the store to private.

This has been done already.

>-Add some e2e tests.

This sounds good.

Am keeping at it. This *seems* doable by tomorrow, though it feels pretty close.

SantosGuillamot commented:
Regarding not triggering the save panel when the post is not published: Although I guess it is not ideal, I believe something like this should work in [this check](https://github.com/WordPress/gutenberg/blob/c0217d1f53a1a743f6a7e8025ab5ab96802533ae/packages/editor/src/components/post-publish-button/index.js#L60) instead of just checking the `hasMetaChanges`:

```
( hasMetaChanges && isPublished ) )
```

We can get the `isPublished` property from the props.

Apart from that, if I understand the logic properly, I think we shouldn't check this [here](https://github.com/WordPress/gutenberg/blob/c0217d1f53a1a743f6a7e8025ab5ab96802533ae/packages/editor/src/components/post-publish-button/label.js#L75).

SantosGuillamot committed:
Suggestion: Move logic to `hasPostMetaChanges` selector

SantosGuillamot commented on packages/editor/src/components/save-publish-panels/index.js:
```diff
@@ -23,6 +24,7 @@ export default function SavePublishPanels( {
+	hasPostMetaChanges,
 	forceIsDirtyPublishPanel,
 } ) {
 	const { closePublishSidebar, togglePublishSidebar } =
```
I think we should get `hasPostMetaChanges` from here and not above, right?

SantosGuillamot commented on packages/editor/src/components/save-publish-panels/index.js:
```diff
@@ -33,6 +35,9 @@ export default function SavePublishPanels( {
 				select( editorStore ).isPublishSidebarOpened(),
 			hasNonPostEntityChanges:
 				select( editorStore ).hasNonPostEntityChanges(),
+			hasPostMetaChanges: unlock(
+				select( editorStore )
+			).hasPostMetaChanges(),
```
I assume we should get `hasPostMetaChanges` from here and not in the component props, right?

SantosGuillamot replied:
I just changed it. Feel free to revert it if you consider it necessary.

SantosGuillamot committed:
Change test formatting

SantosGuillamot committed:
Don't show save panel in pre-publish

SantosGuillamot committed:
Get `hasPostMetaChanges` from the proper place

SantosGuillamot commented on packages/editor/src/components/save-publish-panels/index.js:
```diff
@@ -54,7 +62,7 @@ export default function SavePublishPanels( {
 			/>
 		);
-	} else if ( hasNonPostEntityChanges ) {
+	} else if ( hasNonPostEntityChanges || hasPostMetaChanges ) {
```
To be honest, I don't fully understand what this part of the code does and if it is necessary or not.

gziolo replied:
As far as I remember, this alternative flow triggers when you edit the template of the post. Let me record a screencast to better illustrate it.

gziolo replied:
https://github.com/WordPress/gutenberg/assets/699132/da45473c-9800-449d-aeed-06d4ad91f4d6

gziolo replied:
In effect, what I think we should achieve with this PR would be:
- Regular post – show only Publish
- Regular post with post meta from the same post – show only Publish (covers Footnotes)
  - Follow-enhancement could be highlighting post meta in the Pre-publish screen
- Regular post with any changes to external post - Save + Publish flow

gziolo replied:
To support my line of thinking for the second case, updating the post meta of the currently edited post, isn't a site change. What I see currently after adding footnotes to the published post:

<img width="1197" alt="Screenshot 2024-05-31 at 09 55 38" src="https://github.com/WordPress/gutenberg/assets/699132/bd20d34e-c5bc-4c79-9c0a-8979b22fd38c">

SantosGuillamot replied:
Okay! Thanks a lot for the video 🙂 I believe I have now a better idea of what that part of the code is doing.

If I understand you correctly, your suggestions are aligned with the use cases I stated in [this comment](https://github.com/WordPress/gutenberg/pull/61811#issuecomment-2139757254), right?

gziolo replied:
For the third case, it seems to be pretty close in terms of the flow, but there is no mention that post meta has changed:

<img width="1375" alt="Screenshot 2024-05-31 at 10 11 14" src="https://github.com/WordPress/gutenberg/assets/699132/08c2b45b-14b8-45b3-bdb6-351dd348d8c5">

gziolo replied:
When I also apply modification to the currently edited post, I see that the annotation about post meta gets applied to all modified posts:

<img width="1365" alt="Screenshot 2024-05-31 at 10 14 00" src="https://github.com/WordPress/gutenberg/assets/699132/c25ef49a-87fe-4e59-8b6e-a005d317eabf">

It means that the check for post meta edits gets set to `true` only when the currently edited post is modified, and in other instances incorrect assertion is used.

gziolo replied:
For completeness, this is what I'm seeing in the site editor:

<img width="1371" alt="Screenshot 2024-05-31 at 10 20 55" src="https://github.com/WordPress/gutenberg/assets/699132/75783601-7873-4990-8da0-549b8977a91e">

SantosGuillamot replied:
If I am not mistaken, that is caused because the selector gets the type and id using `getCurrentPost`: [link](https://github.com/WordPress/gutenberg/pull/61811/files#diff-a1cdb9f3f4ce98ee4ac08627daff3c4cb003766468d4bb3780fe0be22a0ad87eR150). Maybe we need to pass the type and id as an argument. I can give that a try.

gziolo replied:
In my opinion, it should use the selector from the `core` store directly so the post id and post type isn't connected with the currently edited entity - post, page, template, etc.

SantosGuillamot replied:
Do you mean using `getEntityRecordNonTransientEdits` directly and removing the `hasPostMetaChanges` one we are creating?

SantosGuillamot replied:
I pushed [this commit](https://github.com/WordPress/gutenberg/pull/61811/commits/baf5ff7973024210315b6991c8cfb1ca474a646e) to illustrate what I mean, but I am happy to iterate once I understand better your suggestion.

If I am not mistaken, that change fixes the problems you mention.

artemiomorales replied:
I tested @SantosGuillamot's fix for to address the scenario wherein the indicator for post meta changes didn't appear when modifying posts in the query loop. It appears to work:

<img width="842" alt="Screenshot 2024-05-31 at 12 45 10 PM" src="https://github.com/WordPress/gutenberg/assets/5360536/a21d8a20-e4a7-44f5-bf30-9b96116d3526">

artemiomorales committed:
Add end-to-end test

artemiomorales committed:
Update class name

artemiomorales committed:
Clarify naming

artemiomorales commented:
@SantosGuillamot Okay, I added the end-to-end test. However, it seems like adding an exception for footnotes the way we were thinking is going to be tricky, or potentially not feasible.

As soon as a footnote is added, all of the registered meta fields show up in the `getEntityRecordNonTransientEdits`, even if they're not even associated with the current post; in this example, it's `text_field` that is not present in the post at all but it still showing up as if it were one of the changes.

<img width="999" alt="Screenshot 2024-05-30 at 10 06 02 PM" src="https://github.com/WordPress/gutenberg/assets/5360536/1ff9369f-45d0-4648-9a73-78074e39410e">

This means that, as far as I can tell, we can't easily distinguish when only a footnote has changed with the current data structures in place. I'll look for another way; in the meantime, additional insight from anyone on this would be much appreciated 🙏

artemiomorales commented:
@SantosGuillamot @gziolo Is circumventing the footnotes a blocker for this PR?

gziolo commented on packages/e2e-test-utils-playwright/src/editor/modify-post-metadata.ts:
```diff
@@ -0,0 +1,44 @@
+import type { Editor } from './index';
+
+/**
+ * Clicks on the button in the header which opens Document Settings sidebar when
```
I don't think this description is correct.

SantosGuillamot replied:
I'm wondering if it is okay to create a new global util or start with something specific for this tests. Something like the [UndoUtils](https://github.com/WordPress/gutenberg/blob/804761ce837eb70b96314b4e12d8fb7fe301dfbe/test/e2e/specs/editor/various/undo.spec.js#L502-L533), for example. (There are many other tests with their own utils)

artemiomorales replied:
I updated the description and moved the test 👍

SantosGuillamot commented:
> Is circumventing the footnotes a blocker for this PR?

I would like to avoid interfering with the current workflows. I'll take a look to see if I can help somehow.

gziolo commented:
> > Is circumventing the footnotes a blocker for this PR?
> 
> In my opinion, I would like to avoid interfering with the current workflows. I'll take a look to see if I can help somehow.

I left my observations from testing this PR in https://github.com/WordPress/gutenberg/pull/61811#discussion_r1621046177. I don't think that the way footnotes block is connected to the post meta is a blocker alone. The way I see it, we need to have a good strategy for when to interfere with the publishing flow and show proper information. The way I would try to approach the initial version would be:
- If the post meta changes apply to the current post in the post editor, only add a section in the pre-publish flow if we agree it is important enough. After publishing the post, we could show the save confirmation, but it's also fine to leave it unchanged until we can show more detailed view of the changes applied that would be helpful for the user.
- If the post meta changes apply to other posts while editing a post in the post editor, then we already surface that, just there information about post meta changes needs to be correctly included.
- In the site editor, we are always in the template editing context, so any changes to post meta in posts (any post type: page, product) should be also marked for consistency.

SantosGuillamot committed:
Show Post Meta in relevant post

SantosGuillamot committed:
Remove extra change

SantosGuillamot commented:
Regarding adding a workaround for `footnotes`, until we figure out it `getEntityRecordEdits` should only return the modified meta fields, I was considering doing something like this:

```js
export const hasPostMetaChanges = createRegistrySelector(
	( select ) => ( state, postType, postId ) => {
		const { type: currentPostType, id: currentPostId } =
			getCurrentPost( state );
		// If no postType or postId is passed, use the current post.
		const edits = select( coreStore ).getEntityRecordNonTransientEdits(
			'postType',
			postType || currentPostType,
			postId || currentPostId
		);

		if ( ! edits?.meta ) {
			return false;
		}

		// Compare if anything apart from `footnotes` has changed.
		const originalPostMeta = select( coreStore ).getEntityRecord(
			'postType',
			postType || currentPostType,
			postId || currentPostId
		)?.meta;

		delete originalPostMeta[ 'footnotes' ];
		delete edits.meta[ 'footnotes' ];

		return ! fastDeepEqual( originalPostMeta, edits.meta );
	}
);
```

`fastDeepEqual` is an external dependency used in `getEntityRecordEdits`: [link](https://github.com/WordPress/gutenberg/blob/cd0ceebf90e4d928f965e6264cc73a6a835161b0/packages/core-data/src/actions.js#L4).

Do you think something like that could make sense? I tested and it seems to work.

artemiomorales committed:
Move test metadata test util

artemiomorales committed:
Update comments

SantosGuillamot commented:
> It seems the hasPostMetaChanges selector is running many times. Is that expected? If I add a console.log, it runs around 500 times

Regarding this, it seems to be happening the same for other selectors like `hasNonPostEntityChanges`, so I assume we don't have to worry too much about it right now?

artemiomorales commented:
> Regarding this, it seems to be happening the same for other selectors like hasNonPostEntityChanges, so I assume we don't have to worry too much about it right now?

I'd say that shouldn't be a blocker for this PR, and we can iterate if need be during the beta period.

artemiomorales committed:
Prevent save panel from appearing when just footnotes are modified

SantosGuillamot reviewed (approved):
Aside from the mentioned follow-ups to improve all the workflows, I think the pull request is good enough 🙂

artemiomorales commented:
I've tested the following scenarios from [@SantosGuillamot's comment](https://github.com/WordPress/gutenberg/pull/61811#issuecomment-2141748869) and [@gziolo's comment](https://github.com/WordPress/gutenberg/pull/61811#discussion_r1621875900), and it all seems to work as expected:

>- When you edit the custom field of a post that is not published: The workflow shouldn’t change (at least from now).
>
>- When you edit the custom field of a post that is published: It should trigger the entities panel showing that the Post Meta has been modified.
>
>- When you edit the custom field of another posts (query loop): It should trigger the entities panel showing that the Post Meta has been modified.
>
>- Regular post with[out] any changes to external post - Save + Publish flow

Regarding the following point, we've added a workaround at least for footnotes thanks to @SantosGuillamot:

>- Regular post with post meta from the same post – show only Publish (covers Footnotes)
Follow-enhancement could be highlighting post meta in the Pre-publish screen

Regarding metadata besides footnotes, the panel appears intentionally when that metadata has changed; there may be other cases similar to footnotes where we don't want to show the save panel, but I believe we can add any additional exceptions during the beta period.

The behavior in the site editor regarding different combinations of modifying metadata for posts also seems to be working as expected as per the third bullet point above. I've also moved the location of the end-to-end test utils around modifying metadata.

That means that the only outstanding points are the following, which I believe will require more investigation and/or discussion to resolve:

>- If I am not mistaken, right now when you edit the custom field, after saving the post, it triggers a message saying “Site updated”.
> 
>- It seems the hasPostMetaChanges selector is running many times. Is that expected? If I add a console.log, it runs around 500 times

I wouldn't consider either of those blockers though and believe this PR is ready to be merged. Thoughts?

artemiomorales committed:
Update package-lock.json

artemiomorales committed:
Revert "Update package-lock.json"

This reverts commit 2b509ce706b02c46fc57e0c3c096d0a5d55b7470.

SantosGuillamot committed:
Update package-lock programatically

artemiomorales commented:
> If I am not mistaken, right now when you edit the custom field, after saving the post, it triggers a message saying “Site updated”.

I created an issue for this bug: https://github.com/WordPress/gutenberg/issues/62236
